### DISCUSSION TITLE
[RFC] `defineMeta` helper function

### DISCUSSION THREAD

**Discussion by AriPerkkio** [storybookjs/storybook#25319] [RFC]
**Status: superseded by https://github.com/storybookjs/storybook/discussions/30112 and implemented**

### Summary

Add a helper function `defineMeta` into Storybook that users can use to type the default export of stories. This helper function will provide type-safety for end-users without needing to write any Typescript at all.

```ts
import { defineMeta } from "@storybook/react";
import { Button } from "./Button";

// Type-safety out-of-the-box without writing any Typescript!
export default defineMeta({
  title: "Example/Button",
  component: Button,
});
```

### Problem Statement

Typing default exports in Typescript cannot be done in a single line that easy. It requires either type assertion, creating a new variable or use of `satisfies` keyword. There's open issue on Typescript project for this: https://github.com/microsoft/TypeScript/issues/13626.
On Storybook's side there's an open issue as well: https://github.com/storybookjs/storybook/issues/18228.

```ts
import { type Meta } from "@storybook/react";
import { Button } from "./Button";

// 1. Requires additional variable 😕 
const meta: Meta = {
  title: "Example/Button",
  component: Button,
};

export default meta;

// 2. Type assertion - not type safe 😢 
export default {
  title: "Example/Button",
  component: Button,
} as Meta;


// 3. `satisfies` keyword. Works well but requires writing Typescript. 😫 
// As far as I know, this is 100% type-safe (compared to #2 above).
export default {
  title: "Example/Button",
  component: Button,
} satisfies Meta;
```

### Non-goals

- Changing any structure of `Meta`
- This RFC considers only helper function of `Meta` typing. In https://github.com/storybookjs/storybook/pull/25069 there's discussion about `defineStory` and `defineConfig` as well, but let's leave those for future to keep scope of this RFC small.

### Implementation

See https://github.com/storybookjs/storybook/pull/25069.

The `defineMeta` helper function itself is very simple. It takes `Meta` as function argument and **returns it as is, without changing it at all**. In runtime code there should be no difference between using `defineMeta` or the `export default { ... }`. This is mostly a change of syntax.

```ts
function defineMeta(meta: Meta): Meta {
  return meta;
}
```

Each framework package of Storybook should export this helper function. By framework packages I mean `@storybook/react`, `@storybook/html` and so on. Example usage in React projects:

```ts
import { defineMeta } from "@storybook/react";

export default defineMeta({});
```

Storybook inspects users' stories by static analysis. Making changes to the supported format requires changes in `csf-tools`. Only the parts where `export default` AST is traversed requires changes.

https://github.com/storybookjs/storybook/blob/e2c05c682cc7c8c7cde358ddb055686513efc257/code/lib/csf-tools/src/CsfFile.ts#L249-L298

In the lines above the analysis is already checking where does the default exported object come from. It accepts the object if it's inlined in `export default`, or if it's coming from variable that is object. This logic needs to be extended to check if:

- `export default` contains `CallExpression` that has name `defineMeta`
- `export default` contains variable. Find that variable and check if it was initialized using `defineMeta` call.

Next analyze the `CallExpression`:

- Check if the first arguemnt of `CallExpression` is `ObjectExpression`
- If it is, use that as Meta in later analysis

### Prior Art

The `defineMeta` is inspired by common conventions used across JS/TS ecosystem. Most modern tools use `defineConfig` that can be used to write type-safe default exports. For example Vite, Playwright, Cypress and Astro use this convention.

### Deliverables

1. Add support to `csf-tools`
2. Introduce `defineMeta` helper in framework packages
3. Update documentation

Future: Introduce more `define...` helper functions.

### Unresolved Questions

- In https://github.com/storybookjs/storybook/pull/25069#issuecomment-1855638854 it was proposed that `defineMeta` would return an object `{ meta, defineStory }`. This would be useful for automatically typing the `defineStory` part. But it would not solve the original issue I've described, as `meta` could not be exported in a single line. I'm not yet sure how `defineStory` could be supported while keeping `export default defineMeta({})` working.

**Comment by tobiasdiez** [comment]
Taking inspiration from how [Vue](https://vuejs.org/api/sfc-script-setup#defineprops-defineemits) works, one could also remove the requirement that the meta is exposed from `export default` (and stories have to be exported as well). It would be enough to just look for the `defineMeta`/`defineStory` function.

Then the following would work:
```ts
import { defineMeta } from '@storybook/react';

import { Button } from './Button';

const { defineStory } = defineMeta({
  component: Button,
});

defineStory({
  name: 'Primary'
  args: {},
});
```

(or one goes even a step further and removes `const { defineStory }` as well, at the cost of adding a language server extension for vscode and other IDEs for proper ts support)

A similar proposal came up in the idiomatic vue implementation: https://github.com/tobiasdiez/storybook-vue-addon/pull/73#discussion_r1321124830

**Comment by shilman** [comment]
This RFC inspired and has been superseded by https://github.com/storybookjs/storybook/discussions/30112, which takes the idea a few steps further. Thanks so much for pushing us in the right direction @AriPerkkio @tobiasdiez !!! 🙏
